### PR TITLE
feat: submit grade

### PR DESCRIPTION
**Pull request by leangseu-edx**
Ticket: [AU-384](https://openedx.atlassian.net/browse/AU-384)
AC:
- [x] Submit will make request to update grade
  - Submit click -> validate grade:
      - if the grade is valid submit current grade
      - else show feedback based on what is missing
- [x] Include submission validation for any missing fields
- [x] Handle success
  - Update the grade. (The grade doesn't persist because I can't change the value on submissions. It shouldn't be a problem with real API)
  - Change grading status to graded
- [x] Unit testing
  - [x]  Component test
  - [ ] Reducer, actions, thunkAction test

Screen Shot 2021-11-09 at 1 35 33 PM

### CODE REVIEW

File: src/data/redux/grading/reducer.js [hunk 1]
```diff
@@ -145,6 +145,27 @@ const grading = createSlice({
     setCriterionFeedback: (state, { payload: { orderNum, value } }) => (
       updateCriterion(state, orderNum, { feedback: value })
     ),
+    validateGrade: (state, { payload: { rubricConfig, gradeData } }) => (
```
Comment (muselesscreator) [inline src/data/redux/grading/reducer.js:148]:
```js
validateGrade: (state, { payload: { rubricConfig, gradeData } }) => {
  const validateFeedback = (value, requirement) =>
    (requirement === feedbackRequirement.required && !value.length);
  return {
    validation: {
      overallFeedback: validateFeedback(rubricConfig.feedback, gradeData.overallFeedback),
      criteria: rubricConfig.criteria.map((criterion, index) => ({
        feedback: validateFeedback(criterion.feedback, gradeData.criteria[index]),
        selected: !gradeData.criteria[index].length,
      }),
    },
  };
},
```

File: src/data/redux/grading/reducer.js [hunk 2]
```diff
@@ -145,6 +145,27 @@ const grading = createSlice({
...
+          feedbackIsInvalid: criterion.feedback === feedbackRequirement.required
+                                && gradeData.criteria[index].feedback?.length === 0,
+          selectedIsInvalid: gradeData.criteria[index].selectedOption?.length === 0,
+        })),
+      })
+    ),
+    completeGrading: (state) => ({
+      ...state,
+      current: {
+        ...state.current,
```
Comment (muselesscreator) [inline src/data/redux/grading/reducer.js:163]: I'm gonna update this in my PR to point to the api response, but this is fine for now.

File: src/data/redux/grading/reducer.js (hunk 1 above)
Comment (leangseu-edx) [inline src/data/redux/grading/reducer.js:148]: That is how you want to structor it?

File: src/data/redux/grading/reducer.js [hunk 3]
```diff
@@ -145,6 +145,27 @@ const grading = createSlice({
     setCriterionFeedback: (state, { payload: { orderNum, value } }) => (
       updateCriterion(state, orderNum, { feedback: value })
     ),
+    validateGrade: (state, { payload: { rubricConfig, gradeData } }) => (
+      updateGradeData(state, {
```
Comment (muselesscreator) [inline src/data/redux/grading/reducer.js:149]: conditional chaining not required.

File: src/testUtils.js [hunk 4]
```diff
@@ -20,14 +22,20 @@ export const formatMessage = (msg, values) => {
...
+  if (typeof contents !== 'object') {
+    return contents;
+  }
   const fn = () => name;
   Object.defineProperty(fn, 'name', { value: name });
-  Object.keys(contents).forEach(nestedName => { fn[nestedName] = contents[nestedName]; });
+  Object.keys(contents).forEach((nestedName) => {
+    const value = contents[nestedName];
+    fn[nestedName] = typeof value !== 'object' ? value
+      : mockNestedComponent(`${name}.${nestedName}`, value);
```
Comment (muselesscreator) [inline src/testUtils.js:38]: yay recursion

File: src/testUtils.js [hunk 5]
```diff
@@ -20,14 +22,20 @@ export const formatMessage = (msg, values) => {
...
-  if (typeof contents !== 'object') { return contents; }
+  if (typeof contents !== 'object') {
+    return contents;
+  }
   const fn = () => name;
   Object.defineProperty(fn, 'name', { value: name });
-  Object.keys(contents).forEach(nestedName => { fn[nestedName] = contents[nestedName]; });
+  Object.keys(contents).forEach((nestedName) => {
+    const value = contents[nestedName];
+    fn[nestedName] = typeof value !== 'object' ? value
```
Comment (muselesscreator) [inline src/testUtils.js]: If the "else"(:) is split onto a new line, I prefer the "then"(?) to be as well for readability.
```suggestion
    fn[nestedName] = typeof value !== 'object'
      ? value
```

File: src/data/redux/grading/selectors.js [hunk 6]
```diff
@@ -169,6 +190,26 @@ selected.criterionFeedback = (state, { orderNum }) => {
...
+};
+
+/**
+ * return criterion selection is invalid
+ * @param {number} orderNum - criterion index
+ * @returns {bool} - criterion feedback is invalid
+ */
+selected.criterionSelectedIsInvalid = (state, { orderNum }) => {
+  const data = module.selected.criterionGradeData(state, { orderNum });
+  return data?.selectedIsInvalid === true;
```
Comment (muselesscreator) [inline src/data/redux/grading/selectors.js:210]: should not have conditional chain.  Should definitely raise an error if you get here and the data is not an object.

Comment (muselesscreator) [reply src/data/redux/grading/selectors.js:210]: same for above

**Review by muselesscreator — CHANGES_REQUESTED** [review]
main request is just to remove optional chaining.

File: src/data/redux/grading/selectors.js [hunk 7]
```diff
@@ -147,6 +147,27 @@ selected.overallFeedback = createSelector(
   (data) => (data ? data.overallFeedback : ''),
 );
 
+/**
+ * Returns rubric-level feedback is invalid
+ * @return {bool} rubric-level feedback is invalid
+ */
+selected.overallFeedbackIsInvalid = createSelector(
+  [module.selected.gradeData],
+  (data) => data?.overallFeedbackIsInvalid === true,
```
Comment (muselesscreator) [inline src/data/redux/grading/selectors.js:156]: this is the one conditional chain that makes sense as rubric feedback component renders/accesses it for ungraded submissions.

File: src/data/redux/grading/reducer.js (hunk 1 above)
Comment (muselesscreator) [reply src/data/redux/grading/reducer.js:148]: kinda, but not in gradeData.  I will worry about in a follow-on PR.

**Review by muselesscreator — APPROVED** [review]

**Pull request merged**